File: leetcode/201-300/228-summary-ranges/bench/summary_ranges.py
"""Benchmark workload for LeetCode #228 — Summary Ranges (Python; scale lane)."""


def summary_metric(nums, n):
    i = 1
    start = nums[0]
    ranges = 0
    esum = 0
    while i <= n:
        if i == n or nums[i] != nums[i - 1] + 1:
            end = nums[i - 1]
            ranges += 1
            esum += start + end
            if i < n:
                start = nums[i]
        i += 1
    return ranges + esum


def main():
    n = 1000000
    passes = 250

    nums = []
    state = 12345
    v = 0
    for _ in range(n):
        state = (state * 1103515245 + 12345) & 2147483647
        v = v + 1 + (state % 3)
        nums.append(v)

    sink = 0
    for _ in range(passes):
        sink += summary_metric(nums, n)
    print(sink)


main()

File: leetcode/201-300/228-summary-ranges/bench/summary_ranges.rs
fn summary_metric(nums: &[i64], n: i64) -> i64 {
    let mut i: i64 = 1;
    let mut start = nums[0];
    let mut ranges: i64 = 0;
    let mut esum: i64 = 0;
    while i <= n {
        if i == n || nums[i as usize] != nums[(i - 1) as usize] + 1 {
            let end = nums[(i - 1) as usize];
            ranges += 1;
            esum += start + end;
            if i < n {
                start = nums[i as usize];
            }
        }
        i += 1;
    }
    ranges + esum
}

fn main() {
    let n: i64 = 1000000;
    let passes: i64 = 250;

    let mut nums: Vec<i64> = Vec::new();
    let mut state: i64 = 12345;
    let mut v: i64 = 0;
    for _ in 0..n {
        state = (state * 1103515245 + 12345) & 2147483647;
        v = v + 1 + (state % 3);
        nums.push(v);
    }

    let mut sink: i64 = 0;
    for _ in 0..passes {
        sink += summary_metric(&nums, n);
    }
    println!("{}", sink);
}
